File: dfeToken.php
<?php 

function createSecureTokenDFe( $password, $uuid ){

  $password = preg_replace("/[^0-9]/", '', $password);
  $key = hash('sha256', $password . ':' . $uuid, true);
  $iv = openssl_random_pseudo_bytes(openssl_cipher_iv_length('AES-256-CBC'));
  $encryptedData = openssl_encrypt(time(), 'AES-256-CBC', $key, OPENSSL_RAW_DATA, $iv);
  $tokenData = json_encode(['data' => base64_encode($encryptedData), 'iv' => base64_encode($iv)]);
  return urlencode(base64_encode($tokenData));

}

$password = '00.000.000/0000-00'; // CPF/CNPJ do tomador
$uuid = '00000000-0000-0000-0000-000000000000'; // UUID da nota fiscal
$key = '00000000000000000000000000000000000000000000'; // Chave da nota fiscal
$url = 'https://nfe.webmaniabr.com/danfe/' . $key . '/?token='.createSecureTokenDFe( $password, $uuid );

?>

File: dfeToken.rb
require 'openssl'
require 'base64'
require 'json'
require 'cgi'

def create_secure_token_dfe(password, uuid)
  password = password.gsub(/\D/, '')
  key_source = "#{password}:#{uuid}"
  key = OpenSSL::Digest::SHA256.digest(key_source)
  iv = OpenSSL::Random.random_bytes(16)
  cipher = OpenSSL::Cipher.new('AES-256-CBC')
  cipher.encrypt
  cipher.key = key
  cipher.iv = iv
  current_time = Time.now.to_i.to_s
  encrypted_data = cipher.update(current_time) + cipher.final
  token_data = {
    data: Base64.strict_encode64(encrypted_data),
    iv: Base64.strict_encode64(iv)
  }
  token_json = token_data.to_json
  token_base64 = Base64.urlsafe_encode64(token_json)
  CGI.escape(token_base64)
end

password = '00.000.000/0000-00'  # CPF/CNPJ do tomador
uuid = '00000000-0000-0000-0000-000000000000'  # UUID da nota fiscal
key = '00000000000000000000000000000000000000000000'  # Chave da nota fiscal
token = create_secure_token_dfe(password, uuid)
url = "https://nfe.webmaniabr.com/danfe/#{key}/?token=#{token}"

puts url
